### ISSUE TITLE
rsyslog loses access to openvpn log file after log rotation when using inotify

### ISSUE THREAD

**Issue by kflavin** [rsyslog/rsyslog#5119]
Hi,
I'm having a problems configuring rsyslog to ship OpenVPN logs to DataDog.  The issue occurs when OpenVPN rotates its log file (OpenVPN handles the rotation itself, it does not use logrotate, see: https://openvpn.net/vpn-server-resources/logging-and-debug-flag-options-for-access-server/#setting-up-log-rotation-for-var-log-openvpnas-log).  

My configuration (included below) works up until the rotation happens, and then I see the following errors message in `systemctl status rsyslog`:

```
imfile: error with inotify API, ignoring file ‘/var/log/openvpnas.log’: Permission denied [v8.32.0]
```

The log file itself is owned by root, but readable by all:
```
ls -l /var/log/openvpnas.log
-rw-r--r-- 1 root root 55759 Apr 18 20:45 /var/log/openvpnas.log
```
Restarting rsyslog causes it to start working again:

```
systemctl restart rsyslog
```

I've also noticed that if I switch the imfile `mode` from inotify to polling, it works after file rotation:
```
# rsyslog.conf
module(load="imfile" mode="polling" PollingInterval="10")
```

So it seems to be some issue with inotify mode.  I'm wondering if anyone has insight into what could be causing this?  OpenVPN creates a new file/inode, it's not truncating the log file in place, so rsyslog seems to be losing access to this new file.  I'd prefer to use inotify mode, but will switch to polling if this is expected behavior and there is no solution.

Rsyslog was configured using instructions here: https://docs.datadoghq.com/integrations/rsyslog/?tab=ubuntuanddebian

### Expected behavior
Rsyslog continues to read from log file and ship logs to DataDog after the log file is rotated.

### Actual behavior
rsyslog is unable to read logs from /var/log/openvpnas.log following log rotation.

### Steps to reproduce the behavior

- Install OpenVPN, and set the log file rotation to something small (like 10000)
- Configure rsyslog as below
- Browse in OpenVPN portal until file rotates

### Environment
- rsyslog version: 8.32.0
- platform: Ubuntu 18.04

/etc/rsyslog.conf: 
```
module(load="imfile" PollingInterval="10")
module(load="imuxsock") # provides support for local system logging
$ModLoad immark
$MarkMessagePeriod 20
module(load="imklog" permitnonkernelfacility="on")
$ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat
$RepeatedMsgReduction on
$FileOwner syslog
$FileGroup adm
$FileCreateMode 0640
$DirCreateMode 0755
$Umask 0022
$PrivDropToUser syslog
$PrivDropToGroup syslog
$WorkDirectory /var/spool/rsyslog
$IncludeConfig /etc/rsyslog.d/*.conf
```

/etc/rsyslog.d/datadog.conf
```
input(type="imfile" ruleset="infiles" Tag="openvpnas" File="/var/log/openvpnas.log")

## Set the Datadog Format to send the logs
$template DatadogFormat,"<api key redacted> <%pri%>%protocol-version% %timestamp:::date-rfc3339% %HOSTNAME% %app-name% - - [metas ddsource=\"openvpn\" ddtags=\"env:prod\"] %msg%\n"

## Define the destination for the logs
$DefaultNetstreamDriverCAFile /etc/ssl/certs/ca-certificates.crt
ruleset(name="infiles") {
    action(type="omfwd" protocol="tcp" target="intake.logs.datadoghq.com" port="10516" template="DatadogFormat" StreamDriver="gtls" StreamDriverMode="1" StreamDriverAuthMode="x509/name" StreamDriverPermittedPeers="*.logs.datadoghq.com" )
}
```

**Comment by davidelang** [comment]
8.32 is very old (about 6 years old now IIRC). The version you are running is 
not actually 8.32, but is 8.32 with some patches that RedHat has backported to 
it. They do not backport all fixes, only ones they consider important enough 
(and we have no idea what they have or have not backported)

As such, we can't support you for a problem like this. There has been a lot of 
work on imfile (including inotify) since 8.32 was released. Please try using the 
community supported version (the latest version was released today, so 8.2304) 
and see if that solves your problem.

David Lang

  On Tue, 18 Apr 2023, 
Kyle Flavin wrote:

> Date: Tue, 18 Apr 2023 13:58:48 -0700
> From: Kyle Flavin ***@***.***>
> Reply-To: rsyslog/rsyslog
>     ***@***.***>
> To: rsyslog/rsyslog ***@***.***>
> Cc: Subscribed ***@***.***>
> Subject: [rsyslog/rsyslog] rsyslog loses access to openvpn log file after log
>     rotation when using inotify (Issue #5119)
> 
> Hi,
> I'm having a problems configuring rsyslog to ship OpenVPN to DataDog.  The issue occurs when OpenVPN rotates its log file (OpenVPN handles the rotation itself, it does not use logrotate, see: https://openvpn.net/vpn-server-resources/logging-and-debug-flag-options-for-access-server/#setting-up-log-rotation-for-var-log-openvpnas-log).
>
> My configuration (included below) works up until the rotation happens, and then I see the following errors message in `systemctl status rsyslog`:
>
> ```
> imfile: error with inotify API, ignoring file ‘/var/log/openvpnas.log’: Permission denied [v8.32.0]
> ```
>
> The log file itself is owned by root, but readable by all:
> ```
> ls -l /var/log/openvpnas.log
> -rw-r--r-- 1 root root 55759 Apr 18 20:45 /var/log/openvpnas.log
> ```
> Restarting rsyslog causes it to start working again:
>
> ```
> systemctl restart rsyslog
> ```
>
> I've also noticed that if I switch the imfile `mode` from inotify to polling, it works after file rotation:
> ```
> # rsyslog.conf
> module(load="imfile" mode="polling" PollingInterval="10")
> ```
>
> So it seems to be some issue with inotify mode.  I'm wondering if anyone has insight into what could be causing this?  OpenVPN is created a new file/inode, it's not truncating the log file in place, so rsyslog seems to be losing access to this new file.  I'd prefer to use inotify mode, but will switch to polling if this is expected behavior and there is no solution.
>
> ### Expected behavior
> Rsyslog continues to read from log file and ship logs to DataDog after the log file is rotated.
>
> ### Actual behavior
> rsyslog is unable to read logs from /var/log/openvpnas.log following log rotation.
>
> ### Steps to reproduce the behavior
>
> - Install OpenVPN, and set the log file rotation to something small (like 10000)
> - Configure rsyslog as below
> - Browse in OpenVPN portal until file rotates
>
> ### Environment
> - rsyslog version: 8.32.0
> - platform: Ubuntu 18.04
>
> /etc/rsyslog.conf:
> ```
> module(load="imfile" PollingInterval="10")
> module(load="imuxsock") # provides support for local system logging
> $ModLoad immark
> $MarkMessagePeriod 20
> module(load="imklog" permitnonkernelfacility="on")
> $ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat
> $RepeatedMsgReduction on
> $FileOwner syslog
> $FileGroup adm
> $FileCreateMode 0640
> $DirCreateMode 0755
> $Umask 0022
> $PrivDropToUser syslog
> $PrivDropToGroup syslog
> $WorkDirectory /var/spool/rsyslog
> $IncludeConfig /etc/rsyslog.d/*.conf
> ```
>
> /etc/rsyslog.d/datadog.conf
> ```
> input(type="imfile" ruleset="infiles" Tag="openvpnas" File="/var/log/openvpnas.log")
>
> ## Set the Datadog Format to send the logs
> $template DatadogFormat,"<api key redacted> <%pri%>%protocol-version% %timestamp:::date-rfc3339% %HOSTNAME% %app-name% - - [metas ddsource=\"openvpn\" ddtags=\"env:prod\"] %msg%\n"
>
> ## Define the destination for the logs
> $DefaultNetstreamDriverCAFile /etc/ssl/certs/ca-certificates.crt
> ruleset(name="infiles") {
>    action(type="omfwd" protocol="tcp" target="intake.logs.datadoghq.com" port="10516" template="DatadogFormat" StreamDriver="gtls" StreamDriverMode="1" StreamDriverAuthMode="x509/name" StreamDriverPermittedPeers="*.logs.datadoghq.com" )
> }
> ```
>
>
> 
>
>
>

**Comment by kflavin** [comment]
Got it.  Thanks.
